xwiki-platform-core/xwiki-platform-menu/xwiki-platform-menu-ui/src/main/resources/Menu/MenuMacro.xml

```diff
@@ -154,28 +154,79 @@ $menuTemplateDoc.content.replace('{', '')
       <cache>long</cache>
     </property>
     <property>
-      <code>require(['jquery'], function($) {
+      <code>define('menu-ui-translation-keys', {
+  prefix: 'menu.ui.',
+  keys: [
+    "openSubMenu",
+    "closeSubMenu"
+  ]
+});
+require(['jquery'], function($) {
   // It's not possible to write a CSS selector that targets list items containing lists so we rely on JavaScript.
   // The 'dropdown' CSS class is used only to display the down/left arrow.
-  $('.menu-horizontal li ul').parent().addClass('xDropdown');
+  $('.menu-horizontal li ul, .menu-vertical li ul').parent().addClass('xDropdown');
 
   // Make sure the menu separators are really empty.
   $('.menu-horizontal, .menu-vertical').find('li &gt; br:first-child').remove();
 
-  // Collapsible menu bahavior.
-  $('.menu-vertical.collapsible').each(function(){
+  // Add aria attributes to the different elements.
+  $('.menu-horizontal').attr('role', 'menubar');
+  $('.menu-vertical').attr('role', 'menu');
+  $('.xDropdown').each(function() {
+    $(this).attr('role', 'menuitem');
+    $(this).attr('aria-haspopup', 'menu');
+  });
+  $('.menu-horizontal, .menu-vertical').find('li')
+    .filter(function(index) {return this.textContent.trim() === ""; })
+    .attr("role","separator");
```

Comment:
I haven't looked in details but it looks like at least some part of this can be done by adding updating `MenuMacro.xml` instead of relying on Javascript.

Reply:
In the last commits I moved the first two of them in the WikiMacroClass velocity code (~l1240).
The others add attributes to nodes inside the `wikimacrocontent`. I think it's relevant to keep them in the javascript extension.